### PR TITLE
add EARLIEST_BY/LATEST_BY to make EARLIEST/LATEST function signatures less ambiguous

### PR DESCRIPTION
**Pull request by clintropolis**
### Description
Follow-up to #11949, this PR splits explicit time column functions into standalone `EARLIEST_BY` and `LATEST_BY` SQL functions to avoid the method signatures being ambiguous and dependent on the column types of the inputs.

Prior to this PR, something like `latest(x, 10)` could either translate to "latest" value of x with max bytes of 10 if a string, but if x was a number it would treat the `10` as a timestamp, so instead of a validation exception like would happen prior to #11949, it would allow it but then explode with strange cast exceptions due to unintended behavior on the users part.

Splitting this out into separate functions makes it much less likely for the user to issue the incorrect query.

<hr>

This PR has:
- [x] been self-reviewed.
- [x] added documentation for new or modified features or behaviors.
- [x] added unit tests or modified existing tests to cover new code paths, ensuring the threshold for [code coverage](https://github.com/apache/druid/blob/master/dev/code-review/code-coverage.md) is met.

### CODE REVIEW

File: docs/querying/sql.md [hunk 1]
```diff
@@ -365,13 +365,13 @@ In the aggregation functions supported by Druid, only `COUNT`, `ARRAY_AGG`, and
...
-|`EARLIEST(expr, maxBytesPerString, timeColumn)`|Like `EARLIEST(expr, timeColumn)`, but for strings. The `maxBytesPerString` parameter determines how much aggregation space to allocate per string. Strings longer than this limit will be truncated. This parameter should be set as low as possible, since high values will lead to wasted memory.|`null` if `druid.generic.useDefaultValueForNull=false`, otherwise `''`|
+|`EARLIEST_BY(expr, timeColumn)`|Returns the earliest value of `expr`, which must be numeric. Earliest value is defined as the value first encountered with the minimum overall value of time column of all values being aggregated.|`null` if `druid.generic.useDefaultValueForNull=false`, otherwise `0`|
+|`EARLIEST_BY(expr, maxBytesPerString, timeColumn)`|Like `EARLIEST_BY(expr, timeColumn)`, but for strings. The `maxBytesPerString` parameter determines how much aggregation space to allocate per string. Strings longer than this limit will be truncated. This parameter should be set as low as possible, since high values will lead to wasted memory.|`null` if `druid.generic.useDefaultValueForNull=false`, otherwise `''`|
```
Comment (imply-cheddar) [inline docs/querying/sql.md:370]: With `EARLIEST_BY` I'd suggest we swap the location of `timeColumn` and `maxBytesPerString`.  Given that `timeColumn` is non-optional on the `_BY` items, we want the optional parameter to be at the end of the signature.

**Review by imply-cheddar — CHANGES_REQUESTED** [review]

File: sql/src/main/java/org/apache/druid/sql/calcite/aggregation/builtin/EarliestLatestBySqlAggregator.java [hunk 2]
```diff
@@ -0,0 +1,180 @@
...
+          StringUtils.format("%s_BY", aggregatorType.name()),
+          null,
+          SqlKind.OTHER_FUNCTION,
+          EARLIEST_LATEST_ARG0_RETURN_TYPE_INFERENCE,
+          InferTypes.RETURN_TYPE,
+          OperandTypes.or(
+              OperandTypes.sequence(
+                  "'" + aggregatorType.name() + "(expr, timeColumn)'\n",
+                  OperandTypes.ANY,
+                  OperandTypes.NUMERIC
```
Comment (gianm) [inline sql/src/main/java/org/apache/druid/sql/calcite/aggregation/builtin/EarliestLatestBySqlAggregator.java:164]: IMO we should allow either NUMERIC or TIMESTAMP here.

**Review by gianm — COMMENTED** [review]
LGTM other than the one comment.

File: sql/src/main/java/org/apache/druid/sql/calcite/aggregation/builtin/EarliestLatestBySqlAggregator.java (hunk 2 above)
Comment (gianm) [reply sql/src/main/java/org/apache/druid/sql/calcite/aggregation/builtin/EarliestLatestBySqlAggregator.java:164]: Hmm, thinking about it more, IMO we should only allow TIMESTAMP. That's consistent with what our other time functions do.

**Review by imply-cheddar — APPROVED** [review]

File: sql/src/main/java/org/apache/druid/sql/calcite/aggregation/builtin/EarliestLatestBySqlAggregator.java (hunk 2 above)
Comment (clintropolis) [reply sql/src/main/java/org/apache/druid/sql/calcite/aggregation/builtin/EarliestLatestBySqlAggregator.java]: changed, and this found a bug when i had to switch to column that could actually be converted to TIMESTAMP (a long), we weren't correctly checking if the time column selector had null values, so I've fixed up the native aggregators to handle this case.

File: .idea/inspectionProfiles/Druid.xml [hunk 3]
```diff
@@ -70,7 +70,6 @@
       </option>
       <option name="IGNORE_FIELDS_USED_IN_MULTIPLE_METHODS" value="true" />
     </inspection_tool>
-    <inspection_tool class="FieldMayBeFinal" enabled="true" level="WARNING" enabled_by_default="true" />
```
Comment (gianm) [inline .idea/inspectionProfiles/Druid.xml:73]: Was this change (& the others in this file) intentional?

Comment (clintropolis) [reply .idea/inspectionProfiles/Druid.xml:73]: no heh, will fix

**Review by gianm — APPROVED** [review]

**Pull request merged**